### ISSUE TITLE
Logging issues with GLOG (composite logging statements don't always work).

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#2731]
We just switched to using GLOG in https://github.com/ray-project/ray/pull/2695.

I'm trying this on MacOS at the current master.

I'm seeing logging statements that log multiple things fail. For example the following works fine.

```cpp
// This works.
RAY_LOG(INFO) << "aaaa";
RAY_LOG(INFO) << 1111;
RAY_LOG(INFO) << 3.14;
RAY_LOG(INFO) << TaskID::from_random();
```

It produces

```
I0823 17:04:43.768009 2788152192 node_manager.cc:1363] aaaa
I0823 17:04:43.768018 2788152192 node_manager.cc:1364] 1111
I0823 17:04:43.768041 2788152192 node_manager.cc:1367] 3.14
I0823 17:04:43.768050 2788152192 node_manager.cc:1368] a7841706c3ecc4f5bfb54b3518eddeb6972bc5fe
```

However, when I do

```cpp
// This does not work.
RAY_LOG(INFO) << "aaaa" << 1111;
RAY_LOG(INFO) << "bbbb" << "cccc";
RAY_LOG(INFO) << 2222 << 3333;
```

I see the following (the second item doesn't appear).

```
I0823 17:04:43.768060 2788152192 node_manager.cc:1369] aaaa
I0823 17:04:43.768069 2788152192 node_manager.cc:1370] bbbb
I0823 17:04:43.768076 2788152192 node_manager.cc:1371] 2222
```

Compare with

```cpp
std::cout << "aaaa" << 1111 << std::endl;;
std::cout << "bbbb" << "cccc" << std::endl;;
std::cout << 2222 << 3333 << std::endl;;
```

which produces the following as desired. 

```
aaaa1111
bbbbcccc
22223333
```

@guoyuhong any idea about this? cc @concretevitamin @pcmoritz

**Comment by atumanov** [comment]
yeah, I'm seeing this as well on MacOSX

**Comment by atumanov** [comment]
same on linux, and disabling it in CMakeLists.txt didn't turn it off:
```
option(RAY_USE_GLOG
  "Build the logging system using glog"
  OFF)
```
commenting it out in src/ray/CMakeLists.txt causes logging to malfunction, printing all log levels, from what it looks like:
```
src/ray/CMakeLists.txt
#if(RAY_USE_GLOG)
#  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DRAY_USE_GLOG")
#  set(RAY_LIB_STATIC_LINK_LIBS ${RAY_LIB_STATIC_LINK_LIBS} ${GLOG_STATIC_LIB})
#  set(RAY_LIB_DEPENDENCIES ${RAY_LIB_DEPENDENCIES} glog)
#endif()
```
Any other workarounds for now?

**Comment by guoyuhong** [comment]
I cannot make the template operator function virtual and the return type is `RayLogBase`, which could be the cause. I will make a fix soon.

**Comment by robertnishihara** [comment]
Fixed by #2734.
